### PR TITLE
Crash copy paste subgraph

### PR DESCRIPTION
**Pull request by julienamsellem**
### Purpose of this PR
https://fogbugz.unity3d.com/f/cases/1364480/

When trying to paste a subgraph node (that is in the clipboard) on that said subgraph it was crashing.
As it does not make sense to support this recursive structure it's now forbidden (the paste option is not allowed in that case).

https://user-images.githubusercontent.com/85612173/143226935-fdd8bf1a-9858-4370-a5c9-0de8ba16e694.mp4

---
### Testing status
Copy+paste same subgraph is correctly forbidden
Copy+paste a different subgraph is still allowed
Copy+paste a simple operator is also allowed

---
### Comments to reviewers
For 2022.x version you must do all operations through the same VFX Window (do not open one window for the main VFX and another one for the subgraph)

### CODE REVIEW

**Comment by VitaSkruibyte** [comment]
Thanks! Issue is no longer reproducible, besides that tried various combinations of reproduction in Subgraph Blocks and Operators. There's one issue that's not caused by the PR but seems to be related:

Editor silently crashes when pasting Subgraph Operator in nested Subgraph Operator. Could this be looked into in this PR or you'd like me to log separate bug for this?

How to reproduce:

1. Create a new graph
2. Create float operator and convert it to Subgraph Operator
3. Convert created operator to Subgraph Operator again
4. Copy the created operator
5. Enter the Subgraph
6. Paste it, Observe that error is thrown
7. Get back to the Parent Graph
8. Copy Subgraph operator again
9. Enter Subgraph Operator and paste it
10. Observe that Unity silently crashed
MBhqGcgv0J

**Comment by VitaSkruibyte** [comment]
Hi, checking changes, initial issue is fixed but when going one step further with Subgraph nesting I still manage to crash the editor.

How to reproduce (100% reproduction): 

1. Create a new graph
2. Create float operator and convert it to Subgraph Operator
3. Convert created operator to Subgraph Operator again
4. Copy the created operator
5. Enter the Subgraph
6. Enter the Subgraph Subgraph
7. Paste it, Observe that error is thrown
8. Get back to the Parent Graph
9. Copy Subgraph operator again
10. Enter Subgraph Operator 
11. Enter Subgraph Subgraph and paste it
12. Observe that Unity silently crashed

Error messages:
```
TLS Allocator ALLOC_TEMP_TLS, underlying allocator ALLOC_TEMP_MAIN has unfreed allocations, size 58
StackOverflowException: The requested operation caused a stack overflow.
```

`UnityEditor.VFX.VisualEffectObjectExtensions.GetResource (UnityEngine.VFX.VisualEffectObject asset) `

`Assertion failed on expression: '!isWriteTxn || s_CurrentSourceAssetTransaction->m_IsWriteTxn'`

Or follow repro in video (might be inconsistent)
bMWGBxBjBy

**Comment by julienamsellem** [comment]
Latest commit should prevent copy/paste a subgraph to any level of its hierarchy

**Review by VitaSkruibyte — APPROVED** [review]
Thank you for looking into it! Issue is fixed and subgraphs no longer lead to Editor crash. Retested several levels of Subgraph Blocks and Operators, parent subgraph can no longer be pasted to the child graph. Additionally tried pasting child subgraphs to parent graphs, didn't notice any outstanding behavior related to PR.

File: com.unity.visualeffectgraph/CHANGELOG.md [hunk 1]
```diff
@@ -14,11 +14,13 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - Creating a new VFX of the same name as an already opened VFX will reuse the existing window [Case 1382841](https://issuetracker.unity3d.com/product/unity/issues/guid/1382841/)
 - Incorrect behavior of Tangent Space in ShaderGraph [Case 1363279](https://issuetracker.unity3d.com/product/unity/issues/guid/1363279/)
 - ShaderGraph made with new VFX SG integration where not listed when searching for a shader graph output [Case 1379523](https://issuetracker.unity3d.com/product/unity/issues/guid/1379523/)
+- Forbid pasting a subgraph in the same subgraph [Case 1364480](https://issuetracker.unity3d.com/product/unity/issues/guid/1364480/)
 
 ## [13.1.2] - 2021-11-05
 
 ### Fixed
 - Removed extra nodes in Ribbon template. [Case 1355602](https://issuetracker.unity3d.com/product/unity/issues/guid/1355602/)
+- Gradient field doesn't support HDR values [Case 1381867](https://issuetracker.unity3d.com/product/unity/issues/guid/1381867/)
```
Comment (PaulDemeulenaere) [inline com.unity.visualeffectgraph/CHANGELOG.md:23]: This is an unexpected change in changelog.md ⏫

File: com.unity.visualeffectgraph/Editor/GraphView/Views/VFXPaste.cs [hunk 2]
```diff
@@ -47,11 +47,52 @@ public static void PasteBlocks(VFXViewController viewController, object data, VF
...
+        static bool CanPasteSubgraph(VisualEffectSubgraph subgraph, string openedAssetPath)
+        {
+            var path = AssetDatabase.GetAssetPath(subgraph);
+            if (path == openedAssetPath)
+            {
+                return false;
+            }
+
+            var resource = VisualEffectResource.GetResourceAtPath(path);
+            var controller = new VFXViewController(resource);
```
Comment (PaulDemeulenaere) [inline com.unity.visualeffectgraph/Editor/GraphView/Views/VFXPaste.cs:59]: It seems really heavy to create a controller on every resources to check the paste subgraph, did you consider to "statify" this behavior ?
By the way, you can access to the list of children of a graph without creating a controller.

**Review by PaulDemeulenaere — COMMENTED** [review]
I have a couple of question pending ⏳

**Review by julienf-unity — APPROVED** [review]

File: com.unity.visualeffectgraph/Editor/GraphView/Views/VFXPaste.cs (hunk 2 above)
Comment (PaulDemeulenaere) [reply com.unity.visualeffectgraph/Editor/GraphView/Views/VFXPaste.cs]: Notice your change here: https://github.com/Unity-Technologies/Graphics/pull/6378/files#diff-08a12858efc2dbffc2c7848ae6458e070543e1d2453fd35a27b0d990a315dcf1R59

**Review by PaulDemeulenaere — APPROVED** [review]
Double check the change, it looks good to me, usage of GetOrCreateGraph will be lighter.
Thanks for the fix.

**Pull request merged**